### PR TITLE
First set of final API updates for 1.0-rc3 release

### PR DESCRIPTION
**Pull request by shefty**
(I expect to have one smaller set of updates to the APIs before 1.0-rc3.)

These changes are best viewed in terms of the individual commits.  The changes center around various updates to the APIs based on OFIWG discussions or open github issues.

* Add mode bit to indicate that remote CQ data consumes a posted receive.  This is needed to support verbs providers.
* Remove FI_REMOTE_SIGNAL flag.  Flag is untested with no strong requirement for use by user space apps.
* Rename FI_REMOTE_COMPLETE -> FI_TRANSMIT_COMPLETE.  Define as default completion mode to make life easier for novice OFI developers..  Definition remains the same, but new name better reflects how it behaves with unreliable data transfers.
* Rename FI_REMOTE_COMMIT -> FI_COMMIT_COMPLETE.  This is left as a placeholder only until a future release.
* Add FI_INJECT_COMPLETE option.  This is the local completion option where a buffer may now be re-used.  Since this is no longer the current default completion mode, a new flag needed to be defined.
* Replace CQ write functionality with CQ signal functionality in order to avoid a performance hit trying to support CQ write.  The EQ write functionality remains.

Note that updates are needed to the providers to support these changes.  I added TODO comments in the provider code to indicate these.  I would like to coordinate having the provider changes available before merging these changes.  Unfortunately, the provider changes are not ones that I can easily do without studying the providers in more detail.

### CODE REVIEW

File: prov/psm/src/psmx_cq.c [hunk 1]
```diff
@@ -732,10 +682,9 @@ static struct fi_ops_cq psmx_cq_ops = {
 	.read = psmx_cq_read,
 	.readfrom = psmx_cq_readfrom,
 	.readerr = psmx_cq_readerr,
-	.write = psmx_cq_write,
-	.writeerr = psmx_cq_writeerr,
 	.sread = psmx_cq_sread,
 	.sreadfrom = psmx_cq_sreadfrom,
+	.signal = fi_no_cq_signal,	/* TODO: write me */
```
Comment (shefty) [inline prov/psm/src/psmx_cq.c]: @j-xiong - I would need help implementing this new call.

File: prov/psm/src/psmx_rma.c [hunk 2]
```diff
@@ -752,7 +752,8 @@ ssize_t _psmx_write(struct fid_ep *ep, const void *buf, size_t len,
 			nargs++;
 		}
 
-		if (flags & FI_REMOTE_COMPLETE) {
+		/* TODO: unsure how to make FI_TRANSMIT_COMPLETE the default */
+		if (flags & FI_TRANSMIT_COMPLETE) {
```
Comment (shefty) [inline prov/psm/src/psmx_rma.c]: @j-xiong - I would need help here as well.

File: prov/sockets/src/sock_cq.c [hunk 3]
```diff
@@ -390,10 +343,9 @@ struct fi_ops_cq sock_cq_ops = {
 	.read = sock_cq_read,
 	.readfrom = sock_cq_readfrom,
 	.readerr = sock_cq_readerr,
-	.write = sock_cq_write,
-	.writeerr = sock_cq_writeerr,
 	.sread = sock_cq_sread,
 	.sreadfrom = sock_cq_sreadfrom,
+	.signal = fi_no_cq_signal,	/* TODO: write me */
```
Comment (shefty) [inline prov/sockets/src/sock_cq.c]: @jithinjosepkl @shantonu - I would need help implementing this

File: prov/sockets/src/sock_progress.c [hunk 4]
```diff
@@ -1680,7 +1674,8 @@ static int sock_pe_progress_tx_send(struct sock_pe *pe,
 		pe_entry->conn->tx_pe_entry = NULL;
 		SOCK_LOG_INFO("Send complete\n");
 		
-		if (!(pe_entry->flags & FI_REMOTE_COMPLETE)) {
+		/* TODO: make FI_TRANSMIT_COMPLETE the default */
+		if (pe_entry->flags & FI_INJECT_COMPLETE) {
```
Comment (shefty) [inline prov/sockets/src/sock_progress.c]: @jithinjosepkl @shantonu - It's not clear to me how to make this the new default.

File: prov/verbs/src/fi_verbs.c [hunk 5]
```diff
@@ -2480,10 +2483,9 @@ static struct fi_ops_cq fi_ibv_cq_data_ops = {
 	.read = fi_ibv_cq_read_data,
 	.readfrom = fi_no_cq_readfrom,
 	.readerr = fi_ibv_cq_readerr,
-	.write = fi_no_cq_write,
-	.writeerr = fi_no_cq_writeerr,
 	.sread = fi_ibv_cq_sread,
 	.sreadfrom = fi_no_cq_sreadfrom,
+	.signal = fi_no_cq_signal,	/* TODO: write me */
```
Comment (shefty) [inline prov/verbs/src/fi_verbs.c]: @a-ilango - I actually could write this function, but I'm hoping that you may be able to help.  :)  @patrickmacarthur needs this functionality for EXS

File: prov/psm/src/psmx_rma.c (hunk 2 above)
Comment (j-xiong) [inline prov/psm/src/psmx_rma.c]: The flag to check here is actually FI_COMMIT_COMPLETE.  The default behavior when flags is 0 does satisfy the transmit complete semantics.

File: prov/psm/src/psmx_cq.c (hunk 1 above)
Comment (j-xiong) [inline prov/psm/src/psmx_cq.c]: Will add this as a separate PR.

**Comment by shefty** [comment]
PR updated with support from Jithin for the sockets provider, and an update to the PSM provider (written as a follow on patch).

**Pull request merged**